src/main/scala/com/exasol/parquetio/reader/RowReadSupport.scala

```diff
@@ -0,0 +1,45 @@
+package com.exasol.parquetio.reader
+
+import com.exasol.parquetio.data.Row
+import com.exasol.parquetio.data.GenericRow
+import com.exasol.parquetio.reader.converter.ParquetRootConverter
+
+import org.apache.hadoop.conf.Configuration
+import org.apache.parquet.hadoop.api.ReadSupport
+import org.apache.parquet.hadoop.api.ReadSupport.ReadContext
+import org.apache.parquet.io.api.RecordMaterializer
+import org.apache.parquet.schema.MessageType
+
+/**
+ * A concrete implementation of [[org.apache.parquet.hadoop.api.ReadSupport]]
+ * that materializes Parquet records into internal [[com.exasol.parquetio.data.Row]]
+ * structure.
+ */
+final class RowReadSupport extends ReadSupport[Row] {
+  override def prepareForRead(
+    conf: Configuration,
+    metadata: java.util.Map[String, String],
+    messageType: MessageType,
+    readContext: ReadContext
+  ): RecordMaterializer[Row] =
+    new RowRecordMaterializer(messageType, readContext)
+
+  override def init(
+    conf: Configuration,
+    metadata: java.util.Map[String, String],
+    messageType: MessageType
+  ): ReadSupport.ReadContext = {
+    val projection = conf.get(ReadSupport.PARQUET_READ_SCHEMA)
+    val requestedSchema = ReadSupport.getSchemaForRead(messageType, projection)
+    new ReadSupport.ReadContext(requestedSchema)
+  }
+
+  private[this] final class RowRecordMaterializer(
+    messageType: MessageType,
+    readContext: ReadContext
+  ) extends RecordMaterializer[Row] {
+    override val getRootConverter: ParquetRootConverter = ParquetRootConverter(messageType)
+    override def skipCurrentRecord(): Unit = getRootConverter.start()
+    override def getCurrentRecord(): Row = new GenericRow(getRootConverter.getResult())
```

Comment:
You could also implement the row just as an accessor on the datastructure. 
So instead of really converting it (and storing the result as object) only implement the interface and convert it on the fly.

But I think your solution is fine. Just wanted to note the idea, since it could be performance optimization.

Reply:
I am not sure if I understood your suggestion. Let us have short call on it.

One reason to get the converted objects into new list is because they are objects are reused for the next records.